evergreen/eg_app/consumers.py

```diff
@@ -0,0 +1,238 @@
+import base64
+import json
+import uuid
+
+# from asgiref.sync import async_to_sync
+from channels.db import database_sync_to_async
+from channels.generic.websocket import AsyncWebsocketConsumer
+
+# from channels.layers import get_channel_layer
+from django.apps import apps
+from django.core.files.base import ContentFile
+
+
+class FeedConsumer(AsyncWebsocketConsumer):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.room_group_name = "feed"
+
+    # connects to socket
+    async def connect(self):
+
+        print("WebSocket connection attempt received!")
+
+        # Add to group
+        await self.channel_layer.group_add(self.room_group_name, self.channel_name)
+        await self.accept()
+        print("(GOOD) WebSocket connection accepted successfully!")
+
+    # disconnects connection handler
+    async def disconnect(self, close_code):
+
+        await self.channel_layer.group_discard(self.room_group_name, self.channel_name)
+        print(f"(BAD) WebSocket disconnected with code: {close_code}")
+
+    # updated to all clients
+    async def feed_update(self, event):
+
+        # sending a lovey message to websocket
+        await self.send(
+            text_data=json.dumps(
+                {"type": "feed_update", "posts": event["data"]["posts"]}
+            )
+        )
+        print("Feed update sent successfully!")
+
+    # handle receiving like events!
+    async def receive(self, text_data):
+        try:
+            data = json.loads(text_data)
+            if data["type"] == "like":
+                post_id = data["post_id"]
+                user = self.scope["user"]
+
+                # Update like in database
+                success, likes = await self.update_like(post_id, user)
+
+                if success:
+                    post_data = await self.get_post_data(post_id)
+
+                    # Broadcast the updated likes to ALL clients
+                    await self.channel_layer.group_send(
+                        self.room_group_name,
+                        {
+                            "type": "like_update",
+                            "post_id": post_id,
+                            "likes": post_data["likes"],
+                            "likers_display": post_data["likers_display"],
+                            "has_liked": post_data["has_liked"],
+                        },
+                    )
+            elif data["type"] == "upload_post":
+
+                post = await self.handle_post_upload(data)
+
+                if post:
+                    # BROADCAST TO ALL CLIENTS
+                    posts_data = await self.get_all_posts()
```

Comment:
Resending every single post will cause all the images to get re-requested. This is extra strain on the server and will probably cause rate limit problems in the demo.

Do it like homework handout where only the new message is sent and appended to the message list

Reply:
I know this is a very inefficient solution, but if I try to do like the homework it refuses to work, and I have "refresh" the browser to see new messages. We can stress test it ourselves before demo. 

I would need some large code refactoring for this, it may need its own branch. For now, it technically works for part 2. My thinking is demo day from present is a greater amount of distance than the deadline for part 2. 

Anyways, it 2:58 AM and I am tired. I'm going to sleep.

Reply:
Fair enough I guess. I'll leave this open for the next meeting though, I don't think it should be that hard to do, especially relative to the benefit

Reply:
Good. In the meantime, I continue to work on it. For now, will resolve so this can be approved and then merged by me.